source/beamline/SGM/energy/SGMEnergyControl.cpp

```diff
@@ -0,0 +1,363 @@
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setUndulatorHarmonic(undulatorHarmonic);
+    }
+}
+
+void SGMEnergyControl::setUndulatorOffset(double undulatorOffset)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setUndulatorOffset(undulatorOffset);
+    }
+}
+
+void SGMEnergyControl::setUndulatorTracking(bool isTracking)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setUndulatorTracking(isTracking);
+    }
+}
+
+void SGMEnergyControl::setGratingTranslationOptimizationMode(SGMEnergyPosition::GratingTranslationOptimizationMode gratingTranslationOptimizationMode)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setGratingTranslationOptimizationMode(gratingTranslationOptimizationMode);
+    }
+}
+
+void SGMEnergyControl::setExitSlitPositionTracking(bool isTracking)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setExitSlitPositionTracking(isTracking);
+    }
+}
+
+void SGMEnergyControl::updateConnected()
+{
+    bool nowConnected = gratingAngleControl_->isConnected() &&
+            gratingTranslationControl_->isConnected() &&
+            undulatorPositionControl_->isConnected() &&
+            exitSlitPositionControl_->isConnected();
+
+    if(nowConnected && !energyPositionController_) {
+
+        initializeEnergyPositionController();
+    }
+
+    setConnected(nowConnected);
+}
+
+void SGMEnergyControl::updateMoving()
+{
+    setIsMoving(gratingAngleControl_->isMoving() ||
+                gratingTranslationControl_->isMoving() ||
+                undulatorPositionControl_->isMoving() ||
+                exitSlitPositionControl_->isMoving());
+}
+
+void SGMEnergyControl::updateValue()
+{
+    // Do this by monitoring the energy controllers energy changed signal
+    // see onEnergyPositionControllerEnergyChanged(double)
+}
+
+void SGMEnergyControl::onGratingTranslationPVValueChanged(double value)
+{
+    SGMGratingSupport::GratingTranslation newGratingTranslation =
+            SGMGratingSupport::encoderStepsToEnum(value);
+
+    if(newGratingTranslation != SGMGratingSupport::UnknownGrating) {
+
+		SGMEnergyPosition::GratingTranslationOptimizationMode savedMode =
+				energyPositionController_->gratingTranslationOptimizationMode();
+
+		energyPositionController_->blockSignals(true);
+		energyPositionController_->setGratingTranslationOptimizationMode(SGMEnergyPosition::ManualMode);
+		energyPositionController_->blockSignals(false);
+
+        energyPositionController_->setGratingTranslation(newGratingTranslation);
+
+		energyPositionController_->blockSignals(true);
+		energyPositionController_->setGratingTranslationOptimizationMode(savedMode);
+		energyPositionController_->blockSignals(false);
+    }
+}
+
+void SGMEnergyControl::onEnergyPositionControllerEnergyChanged(double value)
+{
+    setValue(value);
+}
+
+void SGMEnergyControl::initializeEnergyPositionController()
```

Comment:
Just checking: is the reason that this code in initializeEnergyPositionController() isn't in the constructor is because it should only be executed when SGMEnergyControl is connected for the first time?